lumicks/pylake/population/hmm.py

```diff
@@ -0,0 +1,210 @@
+        return self._model.A
+
+    @property
+    def means(self):
+        """Model state means."""
+        return self._model.mu
+
+    @property
+    def variances(self):
+        """Model state variances."""
+        return 1 / self._model.tau
+
+    @property
+    def std(self) -> np.ndarray:
+        """Model state standard deviations."""
+        return np.sqrt(self.variances)
+
+    def label(self, trace):
+        """Label channel data as states.
+
+        Parameters
+        ----------
+        trace : Slice
+            Channel data to label.
+        """
+        return viterbi(trace.data, self._model)
+
+    def path(self, trace):
+        """Calculate the state and emission paths for a given data trace.
+
+        Parameters
+        ----------
+        trace : Slice
+            Channel data to determine paths.
+
+        Returns
+        -------
+        state_path : Slice
+            Estimated state path
+        emission_path : Slice
+            Estimated emission path
+        """
+        state_path = self.label(trace)
+        emission_path = self.means[state_path]
+
+        state_path = trace._with_data_source(trace._src._with_data(state_path))
+        emission_path = trace._with_data_source(trace._src._with_data(emission_path))
+
+        return state_path, emission_path
+
+    def extract_dwell_times(self, trace, *, exclude_ambiguous_dwells=True):
+        """Calculate lists of dwelltimes for each state in a time-ordered statepath array.
+
+        Parameters
+        ----------
+        trace : Slice
+            Channel data to be analyzed.
+        exclude_ambiguous_dwells : bool
+            Determines whether to exclude dwelltimes which are not exactly determined. If `True`, the first
+            and last dwells are not used in the analysis, since the exact start/stop times of these events are
+            not definitively known.
+
+        Returns
+        -------
+        dict:
+            Dictionary of all dwell times (in seconds) for each state. Keys are state labels.
+        """
+        statepath = self.label(trace)
+        dt_seconds = 1.0 / trace.sample_rate
+
+        dwell_counts, _ = _dwellcounts_from_statepath(
+            statepath, exclude_ambiguous_dwells=exclude_ambiguous_dwells
+        )
+        dwell_times = {key: counts * dt_seconds for key, counts in dwell_counts.items()}
+        return dwell_times
+
+    def pdf(self, x):
+        """Calculate the Probability Distribution Function (PDF) given the independent data array `x`.
+
+        Parameters
+        ----------
+        x : numpy.ndarray
+            Array of independent variable values at which to calculate the PDF.
+
+        Returns
+        -------
+        numpy.ndarray:
+            PDF array split into components for each state with shape (n_states, x.size).
+            The full normalized PDF can be calculated by summing across rows.
+        """
+        return np.exp(self._model.state_log_likelihood(x))
```

Comment:
yeah, I wasn't quite happy with this because technically the HMM doesn't have a PDF in the traditional sense I guess. Considering right now we only use it for the histogram and it's a single line of code, I'm gonna get rid of the method altogether. We can always add it back (private or otherwise) if we find we need it in other places in the future